handlers/stock.js

```diff
@@ -0,0 +1,178 @@
+    command: 'stock',
+    handler: stock
+};
+
+function stock(channel, message, params) {
+
+    // Wont work until this is populated from whatever is managing creds
+    var placeholderConnectionString;
+
+    // Initialize table service
+    const tableSvc = azure.createTableService(placeholderConnectionString);
+    var splitParams = params.split(" ");
+
+    // Verify the Azure Storage Table exists and create if it doesn't
+    tableSvc.createTableIfNotExists(TABLENAME, function (error, result, response) {
+        if (!error) {
+            // Table exists or created
+        }
+    })
+
+    // Verify all users have an account and create new ones for those that don't
+    for (let member of channel.guild.members.array()) {
+        getStock(tableSvc, member.user.username, function (callback) {
+            if (typeof callback != 'number') {
+                console.log("initializing " + member.user.username);
+                updateUser(tableSvc, member.user.username, STARTINGSTOCK);
+            }
+        })
+    }
+
+    if (splitParams[0] == "help") {
+        channel.send("Command Usage:" + "\n" +
+            ".reset - Resets you back to 1000$$" + "\n" +
+            ".balance - Displays your balance" + "\n" +
+            ".allbalances - Displays everyone's balance" + "\n" +
+            ".send [username] [amount] - sends user amount of $$" + "\n");
+    }
+
+    // Reset command to go back to starting stock
+    if (splitParams[0] == "reset") {
+        updateUser(tableSvc, message.author.username, STARTINGSTOCK);
+    }
+
+    // Get the stock for that user
+    if (splitParams[0] == "balance") {
+        getStock(tableSvc, message.author.username, function (callback) {
+            channel.send(message.author.username + "'s balance is " + callback + "$$");
+        });
+    }
+
+    // Show the stock for all users
+    if (splitParams[0] == "allbalances") {
+        showAllStock(tableSvc, channel);
+    }
+
+    // Send stock to another user if they have an account
+    if (splitParams[0] == "send") {
+        var recipient = splitParams[1];
+        var sendAmount = parseInt(splitParams[2]);
+
+        console.log("Send recipient: " + recipient);
+        console.log("Send amount: " + sendAmount);
+
+        if (recipient != message.author.username) {
+            getStock(tableSvc, recipient, function (receiverStock) {
+                console.log(typeof receiverStock)
+                if (typeof receiverStock == 'number') {
+                    getStock(tableSvc, message.author.username, function (senderStock) {
+                        if ((senderStock - sendAmount) > 0) {
+
+                            var newSenderTotal = senderStock - sendAmount;
+                            var newReceiverTotal = receiverStock + sendAmount;
+
+                            console.log("Attempting to send " + sendAmount + " stock");
+
+                            updateUser(tableSvc, message.author.username, newSenderTotal);
```

Comment:
any worry about race conditions or rollbacks?

don't want people creating money from nowhere

Reply:
I'll look into this more later, but the basic logic:
- if getStock fails for either the recipient or sender, neither receiver or sender would be updated
- the sender can't go below 0 stock by sending more than they have

I thought the callbacks protected it from a race but let me know if there's something I'm missing